pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -312,6 +365,62 @@ protected synchronized boolean trySendMessagesToConsumers(ReadType readType, Lis
             }
         }
 
+        // Update the last sent position and remove ranges from individuallySentPositions if necessary
+        if (!allowOutOfOrderDelivery && lastSentPosition != null) {
+            final ManagedLedgerImpl managedLedger = ((ManagedLedgerImpl) cursor.getManagedLedger());
+            com.google.common.collect.Range<PositionImpl> range = individuallySentPositions.firstRange();
+
+            // If the upper bound is before the last sent position, we need to move ahead as these
+            // individuallySentPositions are now irrelevant.
+            if (range != null && range.upperEndpoint().compareTo(lastSentPosition) <= 0) {
+                individuallySentPositions.removeAtMost(lastSentPosition.getLedgerId(),
+                        lastSentPosition.getEntryId());
+                range = individuallySentPositions.firstRange();
+            }
+
+            if (range != null) {
+                // If the lowerBound is ahead of the last sent position,
+                // verify if there are any entries in-between.
+                if (range.lowerEndpoint().compareTo(lastSentPosition) <= 0 || managedLedger
```

Comment:
> * Send `E6` to `c1` and `E9~E11` to `c2`.
>     * State `replay-queue`: `E7, E8`, `sentPostionList`: `[E6, E9~E11]`, `lastSentPistion`: `E5(mark-deleted position)`

In this step, more precisely, the `sentPostionList` is `[(E5, E6], (E8, E11]]`. So, the first range of it is `(E5, E6]`.
The lower endpoint ( `E5` ) is less than or equal to the current lastSentPosition ( `E5` ), so the `lastSentPosition` will be updated to the upper endpoint ( `E6` ).

https://github.com/equanz/pulsar/blob/c3b318af10b45d53c07cc2163a256ce98edf8e8b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L371-L421

> * Add new consumer `c3`, the selector will return `c3` if `key-a`
>     * State `lastSentPosition` of `c3` is `E5`

From the preceding, the `lastSentPosition` of `c3` (joined position) is `E6`.

> * Send nothing to `c3`.

Send `E6` to `c3`.

https://github.com/equanz/pulsar/blob/c3b318af10b45d53c07cc2163a256ce98edf8e8b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L495-L501

Reply:
OK. Correct the flow.

`msgs` on the topic.
- `L: E6` key: `key-a`
- `L: E7` key: `key-a`
- `L: E8` key: `key-a`
- `L: E9` key: `key-b`
- `L: E10` key: `key-b`
- `L: E11` key: `key-b`

---

The flow of sending messages
-  `incoming-queue` of `c1` is `1`, `c2` is `1k` 
- Send `E6` to `c1` and `E9~E11` to `c2`.
  - State `replay-queue`: `E7, E8`, `sentPostionList`: `[E5~E6, E8~E11]`, `lastSentPistion`: `E6(re-calculate after sent)`
- `c1` never acknowledges `E6`.
- Add new consumer `c3`, the selector will return `c3` if `key-a`
  - State `lastSentPosition` of `c3` is `E6` 
- Disconnect c1.
- Next reading: `[E6~E7]`
- Send `E6` to `c3`.
  - State `replay-queue`: `E7, E8`, `sentPostionList`: `[E8~E11]`, `lastSentPistion`: `E6`

Reply:
Just mark this for easier reading: 
- `lastSentPosition` represents the maximum **consecutive** position that has been sent.
- `individuallySentPositions` records the entries that have been sent, helping to calculate the next value of `lastSentPosition`.

Reply:
---

`msgs` on the topic.
- `L: E6` key: `key-a`
- `L: E7` key: `key-a`
- `L: E8` key: `key-a`
- `L: E9` key: `key-b`
- `L: E10` key: `key-b`
- `L: E11` key: `key-b`

---

The flow of sending messages
-  `incoming-queue` of both `c1` and `c2` is `1k` 
- Send `E6~E8` to `c1` and `E9~E11` to `c2`.
  - State `replay-queue`: `[]`, `sentPostionList`: `[]`, `lastSentPistion`: `E11`
- `c1` never acknowledges.
- Add new consumer `c3`, the selector will return `c3` if `key-a`, `incoming-queue` of `c3` is `1` 
  - State `lastSentPosition` of `c3` is `E11` 
- Disconnect `c1`.
- Next reading: `[E6~E8]`
- Send `E6` to `c3`.
  - State `replay-queue`: `E7, E8`, `sentPostionList`: `[E5~E6]`, `lastSentPistion`: `E6(re-calculate after sent)`
- Add new consumer `c4`, the selector will return `c4` if `key-a`, `incoming-queue` of `c4` is `1k` 
  - State `lastSentPosition` of `c4` is `E6` 
- Disconnect `c3`.
- Next reading: `[E6~E8]`
- Send `E6` to `c4`.

Reply:
> * Send `E6` to `c3`.
>     * State `replay-queue`: `E7, E8`, `sentPostionList`: `[E5~E6]`, `lastSentPistion`: `E6(re-calculate after sent)`

The `sentPostionList` and `lastSentPistion` are not cleared by disconnecting a consumer. These are cleared when all consumers are disconnected.
https://github.com/equanz/pulsar/blob/c3b318af10b45d53c07cc2163a256ce98edf8e8b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L173-L179

In this proposal (https://github.com/apache/pulsar/pull/20776) , we care only **not dispatched messages**. We don't care about already dispatched messages.

* [issue-1]: Care about **read but not try to dispatched** messages
* [issue-2]: Care about **skipped** messages

I don't think we need to care about already dispatched messages. Because we can't control client-side redeliver operations from broker-side fully.

simple example:
(Assume that the selector will return `c2`)
```
c1: 3:1, 3:2
c2:
lastSentPosition: 3:2
recentlyJoinedConsumers: {c3: 3:2}
-> (unack 3:2)
c1: 3:1
c2: 3:2
lastSentPosition: 3:2
recentlyJoinedConsumers: {c3: 3:2}
-> (unack 3:1)
c1:
c2: 3:2, 3:1  // out-of-order
lastSentPosition: 3:2
recentlyJoinedConsumers: {c3: 3:2}
```

https://github.com/apache/pulsar/pull/21953#discussion_r1519261144